### ISSUE TITLE
Feature request for completionists

### ISSUE THREAD

**Issue by brabalan** [stom79/mastalab#34]
I really like how Mastalab loads new toots without losing my place in the timeline, but it is currently limited to the 40 newest toots, and I don't know how to easily get the other ones without losing my place.

Could it be possible to either allow a larger number than 40 in the preferences, or to display a button in the timeline saying more toots are available. Tapping on this button would download 40 (or fewer) of the missing toots. Alternatively, there could be a refresh button that does not lose one's place (the only way I know to manually refresh is to go to the top and pull down).

**Comment by stom79** [comment]
Thanks for forwarding your issue.

> Could it be possible to either allow a larger number than 40 in the preferences, or to display a button in the timeline saying more toots are available. 

This value can be up to 80. I will increase it. In home timeline, new toots are retrieved with the streaming API (live events). Instead of a button, there is the counter and only when it is clicked there is a scroll to the new toots.

> Alternatively, there could be a refresh button that does not lose one's place (the only way I know to manually refresh is to go to the top and pull down).

Currently a click on the tab makes two actions: scroll top & display new toots. So, I will display automatically new toots, it will allow to smoothly scroll up by swiping the screen.

**Comment by brabalan** [comment]
Great, thanks a lot!

**Comment by stom79** [comment]
Sorry, I was wrong. For timelines, the limit is [40 toots per answer](https://github.com/tootsuite/documentation/blob/master/Using-the-API/API.md#timelines).

**Comment by brabalan** [comment]
OK, so could it be possible to detect that some toots were not fetched, and add a button to do another API call with ~max_id~ being the oldest toot that was fetched in the last call.

(I suspect that if you do a call with ~since_id~ being the latest toot you know, and there are more than 40 toots to fetch, the one you get are the most recent one, and not the ones following immediately after ~since_id~.)

**Comment by stom79** [comment]
Currently, it works like that:

```
Toots displayed ^      ^-----+ Last toot
when app starts |      |
                | 40   |
                |      |
                |      |
                v      |
                ^      |
Missing toots   |      |                ^  since_id
                |      |                |
                |      |                +
                v      +-----+ Last read toot (bookmark)
                       |                +
                       |                |
                       |                v max_id
                       |
                       |
                       |
                       |
                       |
                       |
                       |        Oldest toots

```
Missing toots are retrieved when scrolling down.
If the focus is put on the `last read toot` when the app starts,  I have to play with `max_id` and `since_id`.

**Comment by stom79** [comment]
And there is the problem of streaming API which will add toots in the timeline even if the top position is not reached.

```
 Streaming API     ^-----+ Last toot
                   |
                   |
⚠ new toot +-----> |
                   |
⚠ new toot +-----> |                last displayed toot
                   |
                   |                ^  since_id
                   |                |
                   |                +
                   +-----+ Last read toot (bookmark)
                   |
                   |
                   |
                   |
                   |
                   |
                   |
                   |
                   |
                   +

```

**Comment by stom79** [comment]
> (I suspect that if you do a call with since_id being the latest toot you know, and there are more than 40 toots to fetch, the one you get are the most recent one, and not the ones following immediately after since_id.)

Yes, you are right. I have no knowledge with Ruby, but pagination looks weird. I had a look at [paginable.rb](https://github.com/tootsuite/mastodon/blob/0a0b9a271a8de84272ff2dafc96ef53db224bb21/app/models/concerns/paginable.rb)

I wonder if this line shouldn't be added to reverse the order when call is done with since_id:

```ruby
module Paginable
  extend ActiveSupport::Concern

  included do
    scope :paginate_by_max_id, ->(limit, max_id = nil, since_id = nil) {
      query = order(arel_table[:id].desc).limit(limit)  if max_id.present?
      query = order(arel_table[:id].asc).limit(limit)  if since_id.present?
      query = query.where(arel_table[:id].lt(max_id)) if max_id.present?
      query = query.where(arel_table[:id].gt(since_id)) if since_id.present?
      query
    }
  end
end

```

**Comment by brabalan** [comment]
I'm surprised by your (very nice) picture: I would have put ~max_id~ toward the top, and ~since_id~ toward the bottom.

Regarding the code you show, although I don't speak ruby, I'm not sure your proposed fix is correct. The current code says: fetch the latest 40 toots, then filter to get those between since and max. Your code says: fetch the latest 40 toots (if max) or the earliest 40 toots (if since), then filter. I don't know what the "earliest" would be, but they might not be what you want. But if it is, you may need to reverse the list to be consistent in both cases.

How about the following: have a button that
- records the current toot id, and the most recent toot id
- asks for 40 toots since the most recent one, but keeps the display on the current toot
- if the most recent one is not included in the fetched toots, display a button « fetch more » to get the missing one, and issue a call with max_id being the oldest of the recently fetched toots
- rinse and repeat

Do you think this could work?

**Comment by stom79** [comment]
From the [API documentation](https://github.com/tootsuite/documentation/blob/master/Using-the-API/API.md#retrieving-a-timeline): 
`max_id`: Get a list of timelines with ID less than this value
`since_id`: Get a list of timelines with ID greater than this value
So by calling max_id, it will retrieve oldest toots (starting from max_id). That's why I put max_id at the bottom and since_id at the top (I don't know if it's clear).

Pagination with `since_id ` should give older toots but with no "hole". Currently, it's not the case. I guess it's due to the order which is the same (DESC). What I expect by ordering them is to move correctly the window:

```
               ^
               |
Last toot +-----+ n+y  +   +----------+
               |       |   |          |   Window calling since_id
               |       |   |    ☹     |
               |       |   |          |      With order DESC
               |       |   |          |
       n+y-40 +-+      +-> +----------+
               |       |
               |       |
               |       |
            ^  |  n+40 |
  page up   |  |       |   +----------+
            |  |       |   |          |
            |  |  n+3  |   |    ☺     |  Window calling since_id
            |  |  n+2  |   |          |
            |  |  n+1  |   |          |      With order ASC
  +---------+  |  n    +-> +----------+
            |  |  n-1
  page down |  |  n+2
            |  |  n-3
            |  |
            |  |
            v  |  n-40
               +

```
So:
`where(arel_table[:id].gt(since_id))`: filter (ie: takes newer toots than `since_id`)
`.order(arel_table[:id])`: move the window
`.limit(limit)`: size of the window

**Comment by brabalan** [comment]
I now understand the diagram, thanks.

To know if there are missing toots, could you check if the current toot is included in the reply? I guess you use `since_id` to avoid fetching know toots, but if you set it to the one before the current one, you could detect that all toots are downloaded by having the current one in the reply.

**Comment by stom79** [comment]
> To know if there are missing toots, could you check if the current toot is included in the reply? 

Yes, it is how it will work:
```
                  ^
                  |
                  +--+  Latest toot  +------------+
                  |                  |            |
                  |                  |   40 last  <-------+  1
                  |                  |   toots    |
                  |                  |            |
                  |   max_id   +--+  +------------+     +   Click on fetch more
                  |                  |            |     |
                  |                  |            |     |
                  |                  |            |     |
                  |                  |            |     |
                  |   max_id'  +--+  +------------+     |   Click on fetch more
                  |                  |            |     |
                  |                  |            |     |
                  |                  |            |     |
                  |                  |            |     |
                  |   max_id'' +--+  +------------+     |   Click on fetch more
                  |                                     |
                  |                                     |
                  |         2                           |
                  |         +                           |
                  |         |        +-------------+    |
                  |  +------v--------+             |    |
Bookmark  +-------+  |Last read toot||             |    |
                     +-----------------------------+    v
                     |               |-|Excluded|--|
            3  +----->  40 toots     +-------------+
                     |   before      |
                     |  bookmark     |
                     +---------------+

```
It needs initially 3 calls:
1 - Fetch first last 40 toots
2 - Fetch only the bookmark toot (since_id/max_id are exclusive)
3 - Fetch 40 toots just before the bookmark

Then the gap between bookmark and first toots will be filled when clicking "Fetch more" which will retrieve 40 oldest toots from the previous window. The "Fetch more button" will be displayed while the bookmark is not retrieved in results. There will be an intersection for the window containing the bookmark to avoid duplicate toots.

**Comment by brabalan** [comment]
Great, thanks!

**Comment by deutrino** [comment]
Is this related to why, when I've been away a while and come back to catch up on a lot traffic, eventually the app just won't load any more?

**Comment by stom79** [comment]
@deutrino No this is not related because this feature is not yet implemented. Could you raise a new issue?

**Comment by stom79** [comment]
OK. Here we go: [This release](https://github.com/stom79/mastalab/blob/e2a5bcf06b66aa72db29c0b8059525b7fa0fb5dd/app/fdroid/release/app-fdroid-release-bookmark.apk?raw=true) should work like describe [here](https://github.com/stom79/mastalab/issues/34#issuecomment-343993447)

**Comment by ve4grm** [comment]
Tried the apk. Everything loaded normally the first time I opened Mastalab, loading new toots at the top (and moving me to the top) like usual. I assume because it didn't have a bookmark.

I closed ML, then opened it a few minutes later to check something. It now loaded a blank Home feed.

If I pulled to refresh, it refreshed the feed, but only to what it was before, minus the new toots it had loaded on the first opening? It weird to describe in words. Basically, it did this:

Open
- NewToot2
- NewToot1
- OldToot1

Close, Reopen
- Empty
- Empty
- Empty

Pull to refresh
- OldToot1
- OldToot2
- OldToot3

I uninstalled and reinstalled from the apk (instead of installing over the previous), same result. I've reverted back to the Play store version for the moment.

**Comment by stom79** [comment]
@ve4grm Thanks for the feedback. Here is [a fix](https://github.com/stom79/mastalab/blob/0eb57c6cc06ea937fb489191d662e1fae62665ee/app/fdroid/release/app-fdroid-release-bookmark-2.apk?raw=true).

**Comment by ve4grm** [comment]
Thanks! I'll test it out and get back to you.

Just a note, when I first opened it, it bright me to the top of the toots I'd already seen, and then loaded a bunch of duplicate toots above it (probably the last 20 or 40?) Before getting to the new ones. Only the actual new toots were marked as new, so that's probably ok if it's a one time thing? 

We'll see how it functions with updates tomorrow!

**Comment by stom79** [comment]
Now it should scroll automatically to the bookmark. Duplicate toots are now discarded.

**Comment by brabalan** [comment]
@stom79 I tried to install it but installation fails (it just tells me « Application non installée » when I try to install it).

**Comment by stom79** [comment]
If the current Fdroid release is installed on your device, it won't update due to the use of different keys (I use my own key and Fdroid its own one). I noticed an issue with duplicate toots. I need to override a method to avoid this issue. So I will publish the new apk soon.

**Comment by brabalan** [comment]
So I need to remove the Fdroid version first? Do I risk losing data? (If it's just a question of entering my password again, it's fine.)

**Comment by ve4grm** [comment]
Alright, so it exhibiting some (I think) unexpected behaviour?

When it tells me there's new toots, and I open it, it loads my home timeline up to the end of the old toots only. Then I have to pull to refresh to see new toots.

This isn't intrinsically a problem, but then when I close it again and come back in, it's still only loaded the (former) old toots, and I have to pull to refresh to see all the toots I'd already seen. My personal preference would be for it to have loaded all of the toots, and just start me scrolled down the timeline to where I'd seen before.

Additionally, it takes a couple seconds to load the (incomplete) home timeline in the first place, which is kind of strange as it didn't before?

And I guess this is only implemented on the home timeline, not the local timeline?

**Comment by stom79** [comment]
OK. Here is [the fix](https://github.com/stom79/mastalab/blob/1051c4a052a43a269e38c9dae13172bba155f34a/app/fdroid/release/app-fdroid-release-bookmark-3.apk?raw=true).

<img src="https://user-images.githubusercontent.com/30171182/33727253-c53fa7a8-db77-11e7-8b75-08ce43a0a324.png" width=400/>

When starting, the app will retrieve the bookmark (last read toot) and will scroll to its position. Toots above are the youngest ones. You need to click on the "fetch more" button to retrieve missing elements (ie: unread toots).

@brabalan 
> So I need to remove the Fdroid version first? Do I risk losing data? (If it's just a question of entering my password again, it's fine.)

Yes, you will lost some data (accounts inside the app will be removed) and a part of your cache.

@ve4grm 
> When it tells me there's new toots, and I open it, it loads my home timeline up to the end of the old toots only. Then I have to pull to refresh to see new toots.

You don't have to pull to refresh (it will reset the bookmark), you only need to click on the "fetch more toots" button. This button is only displayed if there are missing toots. Otherwise just scroll up.

> Additionally, it takes a couple seconds to load the (incomplete) home timeline in the first place, which is kind of strange as it didn't before?

Now it needs two calls to get toots above and below the bookmark. But it should not increase a lot the loading time.

> And I guess this is only implemented on the home timeline, not the local timeline?

Yes. I prefer to start with the home timeline.

**Comment by ve4grm** [comment]
@stom79 

> You don't have to pull to refresh (it will reset the bookmark), you only need to click on the "fetch more toots" button. This button is only displayed if there are missing toots. Otherwise just scroll up.

Right, so that's not happening right now. What I'm getting is:

Open app.
- OldToot1
- OldToot2
- OldToot3

Try to scroll up, can't. Pull to refresh.
- NewToot1
- NewToot2
- OldToot1
- OldToot2

Close app, open again.
- OldToot1
- OldToot2
- OldToot3

Try to scroll up, still can't. Pull to refresh.
- New(but actually now old)Toot1
- New(but actually old)Toot2
- OldToot1
- OldToot2

It feels like it's not completing the call to load the new toots? And then I guess not moving the bookmark?

If I leave it until some NEW new toots come in, then it moves the bookmark up to the top of the OLD new toots, and loads up to there when I open it, now loading the old new toots, but not the new new toots, which I still have to pull to refresh.

That might be confusing. If you need, I can post pictures, or figure out how to record screen capture video?

I'm also going to try uninstalling completely and then installing the apk, see if that fixes anything.

**Comment by stom79** [comment]
@ve4grm These results come from the apk I posted here: https://github.com/stom79/mastalab/issues/34#issuecomment-350031430 ?

**Comment by ve4grm** [comment]
@stom79 Yup. It's weird.

For more info again:
LG G6
Android 7.0

**Comment by ve4grm** [comment]
For additional reference, I just tried:

- clear data/cache
- uninstall
- reinstall from apk 3
- open, see it still remembered my account for some reason, log out
- clear data/cache
- re-login

And it still does the same thing.

**Comment by stom79** [comment]
@ve4grm 

> It feels like it's not completing the call to load the new toots? And then I guess not moving the bookmark?

This new [release](https://github.com/stom79/mastalab/blob/5fd2e9c39c3dc61bcdf8fc7bb714d105fff4263e/app/fdroid/release/app-fdroid-release-bookmark-4.apk?raw=true) should fix this issue.

**Comment by ve4grm** [comment]
Testing now. The first load showed the new toots, but scrolled me to the top (only 2 toots). Maybe because of images loading, or being the first load? I'll wait for new toots to come in and see what it does.

I'm also curious (but may not be able to find out myself, as my feed is pretty slow), if you have like 100 new toots, read 50 and exit in order to come back later, will this system pull you to the top of the old toot stack, or return you to where you were in the stack?

Probably more important to know when it's implemented on local timeline. And federated, I guess. (I don't even use federated timeline, myself.)

**Comment by stom79** [comment]
The app should target the first last toot having the label "new" in blue. If you scroll down of one toot, this one should not have this label because it is supposed to be the bookmark.

>  if you have like 100 new toots, read 50 and exit in order to come back later

Good question, and I will need your help about what you expect. Currently, as soon as you open the app the bookmark is put to the last recent toot. That means if you restart the app (or use pull to refresh) you will lose the current bookmark. 

> (I don't even use federated timeline, myself.)

You can disable and hide this timeline in settings :)

**Comment by ve4grm** [comment]
> The app should target the first last toot having the label "new" in blue. If you scroll down of one toot, this one should not have this label because it is supposed to be the bookmark.

Ok. What should I expect to see when I open the app, though?

1. Only old toots on screen, and scroll up to see the new ones?
2. The first new toot at the top of the screen?
3. The screen full of new toots, and I have to scroll down to see old ones?

All are probably fine, but I'd like to calibrate my expectations, so I can tell if it's behaving or not.

> > if you have like 100 new toots, read 50 and exit in order to come back later

> Good question, and I will need your help about what you expect. Currently, as soon as you open the app the bookmark is put to the last recent toot. That means if you restart the app (or use pull to refresh) you will lose the current bookmark.

Well, my ideal (though I won't speak for others) would be that it keeps your place in the feed, no matter where you leave off. I guess that would translate to:

IF (toot has been shown on the screen) AND (a more recent toot isn't already bookmarked)
THEN (mark toot with bookmark)

That said, I do get that there might be technical limitations for that?

I figure that if this was the behaviour, it should work for most use cases. There's always a way to get to the top of the feed, thereby setting the bookmark to the top toot, after all.

> You can disable and hide this timeline in settings :)

No worries, I do. :-)

**Comment by ve4grm** [comment]
Alright, so it seems like it's still having the same issues? Load Mastalab, see old toots, have to pull to see new toots, close and open mastalab, see same old toots, still have to pull to see the same "new" ones that I had to pull for before.

This is weird...

**Comment by ve4grm** [comment]
Ok, more testing info.

I got a notification saying "New post from <person>, plus 4 more toots" - so total 5, right?

I opened the app, and saw... 3 new toots.
(I think it was scrolled to the top of them, but am unsure. Regardless, all three were on the screen.)

I pulled to refresh, and got an additional new toot, for a total of 4.

Exit app, re-open. I see the 3 toots I saw before, and can still pull to refresh to see the final one.

So it isn't hiding all the toots? But it won't load the last one without a pull/refresh. (The last post, I think I only had one new toot.)

**Comment by ve4grm** [comment]
And... this time, I opened the app, it showed me the one new toot (beyond the previous "new" ones) at the top like it should, and there was nothing to pull to refresh.

**Comment by brabalan** [comment]
I'm giving a try of the latest release, and here are my observations
- new toots appears without having to refresh, which is great (this fixes the bug I reported on Mastodon)
- when I'm at the top of the timeline and I later go back to the app, I'm still at the top, so I'm seeing the newest toot. I would like to keep my place, i.e., see the toot I was seeing before.
- if I'm not at the top of the timeline, I also don't keep my place, in the sense that new toots that appear push down the old ones, so I no longer see the old one.

It seems like the way things currently work is that the position relative to the top is preserved, whereas I would prefer to have the toot currently displayed preserved.

Now I need to get off Mastodon for long enough to test the "load more toots" feature ;)

**Comment by ve4grm** [comment]
Right, so I opened up mastodon this morning. It loaded with the first "new" toot at the top of the screen, and allowed me to scroll up to see other new toots. When I got to the top, and pulled to refresh, nothing additional showed up.

In other words, it did what I would have expected. So that's weird? 

My guess is the difference is that all these new toots came during different update cycles during the night, compared to before where they were brand new from the latest cycle? 

I'll keep monitoring.

**Comment by stom79** [comment]
Toots at the top are the 40 most recent ones. Then it's your bookmark. At the bottom of the last most recent toots (ie number 40) there is a fetch more button to display missing toots. As soon as you pulled to refresh you lost the bookmark.
In the next release the bookmark will be put automatically to the the toot having the focus.

**Comment by ve4grm** [comment]
And with the new toots that just came in, it's still hiding the most recent toot behind the pull to refresh. I'm going to try to capture some video to show what's happening.

**Comment by stom79** [comment]
Pull to refresh is useful when a bug occurs with refresh. As soon as you get a new toot, the counter is increased and by clicking on the tab you scroll up to the new toots.

**Comment by ve4grm** [comment]
Right. Gifs attached for clarity.

Gif 1 - At this point, there was no notification. When I opened it, there were a couple new toots. Both the ons from Felyn were new. One showed, one had to be pulled to refresh. After I do so, I close and reopen the app so you can see it stays like that, and has to be pulled again.
![gpscreen1](https://user-images.githubusercontent.com/18155499/33775035-bcdd3284-dc02-11e7-997b-f98b46a94372.gif)

Gif 2 - Notification of new toots. When I open it, the top toot is indeed new, but isn't marked as such. When I pull to refresh, another toot appears. It isn't shown, but as above that toot stays hidden if I exit and re-enter. Also, opening the app didn't clear the notification of new toots?
![gpscreen2](https://user-images.githubusercontent.com/18155499/33775038-c284d05c-dc02-11e7-995d-956a8a9bc8f4.gif)

Gif 3 - Here it seems to be working properly maybe? I had sent some toots from the web interface. I got a notification that I had been replied to, though it was by someone who didn't follow me so it didn't show up on my home feed. When I load the app, it starts me at the bottom of the new toots, all of which happen to be mine, and lets me scroll to the top. Nothing appears when I pull to refresh.

My thought is that my toots were loaded in a previous load cycle, but didn't generate a "new toot" notification because they were by me? It only seems to mess up on the toots from the current load cycle. The one other time it worked, I still had new toots from a prior cycle as well.
![gpscreen3](https://user-images.githubusercontent.com/18155499/33775050-c6eef532-dc02-11e7-95b0-710c43ea48f8.gif)

I hope all of this helps.

**Comment by stom79** [comment]
> At this point, there was no notification. When I opened it, there were a couple new toots. 

Notifications for new toots are not in live  (it's a service that collects new toots every 30 minutes). But it might have an issue because if toots have been read with the app, they should not be sent to the notification service.

GIF 1: There is no toot flagged as new (even before the pull to refresh). You closed the app before the test, unfortunately you reset the bookmark to the newest toot and all toots are flagged as read (with this version).


Gif 2 - I will see the issue with new toots and notifications


Gif 3:
> but didn't generate a "new toot" notification because they were by me?

Yes.

>  Nothing appears when I pull to refresh.

Toots are displayed but pull to refresh reset the bookmark (at the newest toot) so there is no issue with this behavior.

**Comment by ve4grm** [comment]
> Notifications for new toots are not in live (it's a service that collects new toots every 30 minutes). But it might have an issue because if toots have been read with the app, they should not be sent to the notification service.

No, no problem with this. I was just trying to clarify the situation, in case it mattered. It didn't later notify me about them or anything.

> GIF 1: There is no toot flagged as new (even before the pull to refresh). You closed the app before the test, unfortunately you reset the bookmark to the newest toot and all toots are flagged as read (with this version).

The flagging as new did work here, I think. I believe they were marked as new before I started recording, but then I read them. So I don't think this was an error with the flagging here. Just in gif 2.

> Toots are displayed but pull to refresh reset the bookmark (at the newest toot) so there is no issue with this behavior.

Yep, no worries. Again, just describing what happened, and that it was different from the other time. Everything about gif 3 was how I expect gifs 1 and 2 should have acted.

**Comment by ve4grm** [comment]
Ok, new weird occurrence. Maybe it helps you solve this, maybe just makes it more confusing.

Got a new toot notification. Clicked it. Mastalab opened to the bottom of three toots marked as new. Perfect! 

Except, when I scrolled down, there were 5 more toots I hadn't seen yet, none of which were marked new.

Don't know if this is related at all.

**Comment by stom79** [comment]
Here is [the last apk](https://github.com/stom79/mastalab/blob/945cbe5c928c26541ab5fcca4dfd1b0293d179b5/app/fdroid/release/app-fdroid-release-bookmark-5.apk?raw=true). It will remember your previous position.

**Comment by ve4grm** [comment]
I'll check it out. Thanks.

**Comment by stom79** [comment]
OK. Sorry, but you should use [this apk](https://github.com/stom79/mastalab/blob/0cf244dafdab78050f4f9965c30ca34793b16e93/app/fdroid/release/app-fdroid-release-bookmark-5.apk?raw=true) which fixes a small issue.

**Comment by ve4grm** [comment]
So, one issue so far. When I go to the local timeline for the first time, it sits on the loading animation forever, until I pull refresh.

Home timeline loading is also extended, but only for a second or so.

I think maybe local timeline is looking for a bookmark but it doesn't exist yet? 

And just to confirm, I am using the latest one with the fix from your last comment.

**Comment by stom79** [comment]
@ve4grm It has been fixed :) You can update the app on Google Play or use the last apk here: https://github.com/stom79/mastalab/releases/tag/1.6.5.1

I will close this issue because it's too long and it becomes hard to follow it :)

**Comment by brabalan** [comment]
Thanks a lot for your great work!
